### PR TITLE
refactor: Migrate ItemAnnotator and EntityLabelChangedEvent to StyledText

### PR DESCRIPTION
**Pull request by Mcrtin**
someone has to check the Patterns

### CODE REVIEW

**Review by kristofbolyai — COMMENTED** [review]
Did you not test it or you couldn't figure this out?

**Comment by Mcrtin** [comment]
i already did all changes and just want to introduce them slowly and I also missed 2 or 3 classes

**Comment by kristofbolyai** [comment]
Then what does "someone has to check patterns" mean?

**Comment by Mcrtin** [comment]
I've marked every pattern where I'm not sure whether I removed all escapes or not

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 1]
```diff
@@ -105,6 +107,10 @@ public static StyledText fromComponent(Component component) {
         return new StyledText(parts, temporaryWorkaround, clickEvents, hoverEvents);
     }
 
+    public static StyledText fromComponent(Component component, boolean onlyCodes) {
+        return onlyCodes ? fromString(component.getString()) : fromComponent(component);
+    }
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java]: What is this and why is this needed? I don't event think this properly works. Either way, we should not need to use this and if there are places where we depend on only using color codes from components and not actual styles, then we need to update those places to be correct.

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 2]
```diff
@@ -114,6 +120,10 @@ public static StyledText fromCodedString(CodedString codedString) {
         return fromString(codedString.getInternalCodedStringRepresentation());
     }
 
+    public String getString() {
+        return getString(PartStyle.StyleType.DEFAULT);
+    }
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java:129]: We didn't need to add this or any change related to adding this. As the comment states on the original `getString` method, in an ideal world that method would be private..

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 3]
```diff
@@ -355,6 +373,30 @@ int addHoverEvent(HoverEvent hoverEvent) {
         return hoverEvents.size();
     }
 
+    public int getWidth() {
+        return McUtils.mc().font.width(getString(PartStyle.StyleType.NONE));
+    }
+
+    public StyledText[] split(String s) {
+        return Arrays.stream(getString().split(s)).map(StyledText::fromString).toArray(StyledText[]::new);
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java]: This is not even used. Plus, this is a dangerous operation which might not be the way to go about things.

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 4]
```diff
@@ -355,6 +373,30 @@ int addHoverEvent(HoverEvent hoverEvent) {
...
+
+    public StyledText subtext(int start, int end) {
+        return StyledText.fromString(getString().substring(start, end));
+    }
+
+    public StyledText subtext(int start) {
+        return StyledText.fromString(getString().substring(start));
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java]: I don't like these kinds of operations in StyledText, just like `split`, this is not the way to do things. Let's remove this and find an alternative way for their usages.

**Review by kristofbolyai — COMMENTED** [review]
I did not get to review everything, here is my comments so far.

File: common/src/main/java/com/wynntils/core/text/StyledText.java (hunk 3 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/core/text/StyledText.java]: This is used in some places i haven't pushed yet. e.g.
```java
private long calculateMessageReadTime(List<StyledText> msg) {
        int words = StyledText.join(" ", msg).split(" ").length;
        long delay =
                dialogAutoProgressDefaultTime.get() + ((long) words * dialogAutoProgressAdditionalTimePerWord.get());
        return delay;
    }
```
or
```java
StyledText[] colored = description.split("\n");
```

I just realized why this is so dangerous...

File: common/src/main/java/com/wynntils/core/text/StyledText.java (hunk 1 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/core/text/StyledText.java]: this is the replacement of fromComponentIgnoringComponentStylesAndJustUsingFormattingCodes as it has such a long name I thought it must be very important ^^. i just copied the behavior to not introduce new bugs.

File: common/src/main/java/com/wynntils/core/text/StyledText.java (hunk 2 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/core/text/StyledText.java:129]: so, should i move the logic to StyedText if possible?
what should I do here:
```java
WynntilsMod.info("Message Edited: " + msgContainer.getRenderTask() + " -> " + newMessage.getString());
```
wrap everything in StyledText?

File: common/src/main/java/com/wynntils/core/text/StyledText.java (hunk 4 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/core/text/StyledText.java]: what do we do here then:
```java
private StyledText getBaseName(StyledText name) {
        int bracketIndex = name.getString().lastIndexOf('[');
        return bracketIndex == -1 ? name : name.subtext(0, bracketIndex);
    }
```

File: common/src/main/java/com/wynntils/core/text/StyledText.java (hunk 3 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/core/text/StyledText.java]: In these tricky cases just use getString and I will eventually fix them up later in the codebase.

File: common/src/main/java/com/wynntils/core/text/StyledText.java (hunk 4 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/core/text/StyledText.java]: This can easily implemented with StyledText. Most likely the `Item Name [97%]` will be 2 parts: `Item Name` and `[97%]`. I just have to think about how to not make it look weird (we would have to use iterateBackwards as of right now).

Again, use getString and I will worry about it later.

File: common/src/main/java/com/wynntils/core/text/PartStyle.java [hunk 5]
```diff
@@ -15,6 +15,9 @@
 import net.minecraft.network.chat.Style;
 
 public final class PartStyle {
+
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/PartStyle.java:18]:
```suggestion
```

File: common/src/main/java/com/wynntils/core/text/PartStyle.java [hunk 6]
```diff
@@ -15,6 +15,9 @@
...
+    public static final PartStyle NONE =
+            new PartStyle(StyledTextPart.EMPTY, CustomColor.NONE, false, false, false, false, false, null, null);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/PartStyle.java:20]:
```suggestion
    public static final PartStyle NONE =
            new PartStyle(StyledTextPart.EMPTY, CustomColor.NONE, false, false, false, false, false, null, null);
            
```

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 7]
```diff
@@ -355,6 +377,51 @@ int addHoverEvent(HoverEvent hoverEvent) {
...
+            return Optional.of(JsonParser.parseString(json).getAsJsonObject());
+        } catch (JsonSyntaxException e) {
+            return Optional.empty();
+        }
+    }
+
+    public PartStyle getStyleAt(int index) {
+        for (StyledTextPart part : parts) if ((index -= part.length()) < 0) return part.getPartStyle();
+        return PartStyle.NONE;
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java:395]: This is logically wrong I think, but I also don't see any case where this should be used. I checked the usages and I think I can find a better solution for this later

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 8]
```diff
@@ -355,6 +377,51 @@ int addHoverEvent(HoverEvent hoverEvent) {
...
+
+    public StyledText subtext(
+            int start, int end) { // todo: doesn't work with color codes because the parts don't keep their color
+        return StyledText.fromString(getString().substring(start, end));
+    }
+
+    public StyledText subtext(
+            int start) { // todo: doesn't work with color codes because the parts don't keep their color
+        return StyledText.fromString(getString().substring(start));
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java]: The todo is part of the reason why this need to be removed...

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 9]
```diff
@@ -355,6 +377,51 @@ int addHoverEvent(HoverEvent hoverEvent) {
...
+
+    public StyledText replaceAll(String regex, String replacement) {
+        return StyledText.fromString(getString().replaceAll(regex, replacement));
+    }
+
+    public StyledText replace(String target, String replacement) {
+        return StyledText.fromString(getString().replace(target, replacement));
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java]: Implement these in a way that you can only replace text parts, not formatting codes. You also need to do this for every part and construct a new styled text that way, because this way you lose some crucial info.

File: common/src/main/java/com/wynntils/core/text/StyledText.java [hunk 10]
```diff
@@ -355,6 +377,51 @@ int addHoverEvent(HoverEvent hoverEvent) {
...
+
+    public Optional<JsonObject> toJsonObject() {
+        String json = StringUtils.substringBeforeLast(getString(), "}") + "}"; // remove extra unnecessary info
+        try {
+            return Optional.of(JsonParser.parseString(json).getAsJsonObject());
+        } catch (JsonSyntaxException e) {
+            return Optional.empty();
+        }
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledText.java:390]: What is even this and why is it needed?

File: common/src/main/java/com/wynntils/core/text/StyledTextPart.java [hunk 11]
```diff
@@ -159,6 +161,14 @@ boolean isBlank() {
         return text.isBlank();
     }
 
+    public int length() {
+        return text.length();
+    }
+
+    public int width() {
+        return McUtils.mc().font.width(text);
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/text/StyledTextPart.java]: Remove everything that has to do with width from StyledText.
Instead create a method in `FontRenderer` where you can pass StyledText and get its width.

File: common/src/main/java/com/wynntils/features/inventory/BulkBuyFeature.java [hunk 12]
```diff
@@ -39,8 +37,7 @@ public class BulkBuyFeature extends Feature {
     @RegisterConfig
     public final Config<Integer> bulkBuyAmount = new Config<>(4);
 
-    // Test suite: https://regexr.com/7998g
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/inventory/BulkBuyFeature.java:42]: Why did you remove this..?

File: common/src/main/java/com/wynntils/handlers/item/ItemAnnotator.java [hunk 13]
```diff
@@ -4,10 +4,10 @@
...
 package com.wynntils.handlers.item;
 
-import com.wynntils.core.text.CodedString;
+import com.wynntils.core.text.StyledText;
 import net.minecraft.world.item.ItemStack;
 
 @FunctionalInterface
 public interface ItemAnnotator {
-    ItemAnnotation getAnnotation(ItemStack itemStack, CodedString name);
+    ItemAnnotation getAnnotation(ItemStack itemStack, StyledText name);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/handlers/item/ItemAnnotator.java:12]: This single change was big enough for a PR. (For future reference).

File: common/src/main/java/com/wynntils/core/text/StyledText.java (hunk 10 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/core/text/StyledText.java:390]: this is from LoreUtils. but due to this being only used once we can also drop the change

File: common/src/main/java/com/wynntils/features/inventory/BulkBuyFeature.java (hunk 12 above)
Comment (Mcrtin) [reply common/src/main/java/com/wynntils/features/inventory/BulkBuyFeature.java:42]: i changed the pattern
